### PR TITLE
Fix TestAutograd.test_pinverse not actually testing

### PR DESCRIPTION
**Pull request by ssnl**
cc @vishwakftw 

Also added a check if none of the input tensors in `gradcheck` have `requires_grad=True`.

### CODE REVIEW

**Comment by vishwakftw** [comment]
Oops, I am so sorry about this. I thought `gradcheck` converted the tensors to `requires_grad=True`.

**Review by vishwakftw — APPROVED** [review]
ONNX changes have come with this PR.

**Comment by SsnL** [comment]
@vishwakftw nah, it's mostly my fault. having worked with gradcheck for 8 months and still forget about this.

**Comment by SsnL** [comment]
Thanks for the reminder! Fixed onnx.

**Review by soumith — APPROVED** [review]

**Comment by SsnL** [comment]
cc @fritzo @fehiepsi @apaszke on `test_distributions` changes. 

In this patch, I added an error if none of the input args is a tensor that requires grad. In `TestDistributions`, we are grad checking some constructors on argument lists with no tensors. So I had to add samples as a grad checked input. Hope that this change is okay with you guys.

**Review by fritzo — APPROVED** [review]
distributions changes LGTM

**Comment by SsnL** [comment]
cc @t-vi `gradcheck` on `einsum` wasn't enabled. After fixing in [this commit](https://github.com/pytorch/pytorch/pull/9192/commits/96ce9f7c5dfcc8f98c048db6c09584fb14f2c2d5), there are three cases that fail the check. Do you happen to have some idea why it is happening? I will look into this in a couple days.

**Comment by t-vi** [comment]
@SsnL Ha, thanks for catching this! I think this is as simple as having double:
With
```
                gradcheck_inps = tuple(t.detach().requires_grad_() for t in test[1:])
```
passses all three cases that were failing before.
An alternative is, of course, to use just add dtype=torch.double for the generated args.
Do you want me to send a diff or include it in yours?

**Comment by SsnL** [comment]
@t-vi I remember that they fail with a 2x Jacobian difference, but I'll have to try again.

**Comment by SsnL** [comment]
oh okay I know why. it's not a bug. it's just that these tests use multiples of same tensor, so numerical perturbation changes not one, but two tensors at the same time!

**Comment by SsnL** [comment]
i'll fix that

**Comment by SsnL** [comment]
No.. even if `detach()` is called, they still share storage.

**Comment by t-vi** [comment]
Ah, but the double() created a separate instance for me...

**Comment by SsnL** [comment]
@t-vi This line is probably the culprit:
https://github.com/pytorch/pytorch/blob/84884dc2d36b278e02a7e50aa26e2d3a2ca56901/test/common.py#L36

:P
